File: fetch_osm.py
import urllib.request
import urllib.parse
import json
import time

query = """
[out:json][timeout:90];
area["name"="Portugal"]->.searchArea;
(
  node["amenity"="charging_station"](area.searchArea);
  way["amenity"="charging_station"](area.searchArea);
  relation["amenity"="charging_station"](area.searchArea);
);
out center;
"""

url = 'https://overpass-api.de/api/interpreter'
params = urllib.parse.urlencode({'data': query.strip()}).encode('utf-8')
req = urllib.request.Request(url, data=params, headers={
    'User-Agent': 'AntigravityEVDataFetcher/1.0 (https://example.com/)'
})

print("Fetching charging stations for all of Portugal from OSM... This may take a minute.")
try:
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode('utf-8'))
        elements = data.get('elements', [])
        print(f"Found {len(elements)} stations in OSM")
        
        # Convert way/relation centers to lat/lon for easy processing
        for element in elements:
            if 'center' in element:
                element['lat'] = element['center']['lat']
                element['lon'] = element['center']['lon']
                
        with open('osm_data.json', 'w', encoding='utf-8') as f:
            json.dump(elements, f, ensure_ascii=False, indent=2)
        print("Data saved to osm_data.json")
except Exception as e:
    print(f"Error: {e}")

File: fetch_osm.js
const fs = require('fs');
const path = require('path');

const query = `
[out:json][timeout:90];
area["name"="Portugal"]->.searchArea;
(
  node["amenity"="charging_station"](area.searchArea);
  way["amenity"="charging_station"](area.searchArea);
  relation["amenity"="charging_station"](area.searchArea);
);
out center;
`;

async function main() {
  const url = 'https://overpass-api.de/api/interpreter';
  console.log('Fetching charging stations for all of Portugal from OSM... This may take a minute.');

  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers: {
        'User-Agent': 'FuelSmartPortugal/2.0',
        'Accept': 'application/json'
      },
      body: new URLSearchParams({ data: query.trim() })
    });

    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const data = await resp.json();
    const elements = data.elements || [];
    console.log(`Found ${elements.length} stations in OSM`);

    for (const el of elements) {
      if (el.center) {
        el.lat = el.center.lat;
        el.lon = el.center.lon;
      }
    }

    const outPath = path.join(__dirname, 'osm_data.json');
    fs.writeFileSync(outPath, JSON.stringify(elements, null, 2), 'utf-8');
    console.log('Data saved to osm_data.json');
  } catch (err) {
    console.error('Error:', err.message);
    process.exit(1);
  }
}

main();
